lib/std/Build/Step/Run.zig

```diff
@@ -588,12 +584,81 @@ fn make(step: *Step, prog_node: *std.Progress.Node) !void {
         argv_list.items[placeholder.index] = cli_arg;
     }
 
-    try runCommand(self, argv_list.items, has_side_effects, &digest, prog_node);
+    try runCommand(self, argv_list.items, has_side_effects, tmp_dir_path, prog_node);
 
     if (self.dep_output_file) |dep_output_file|
         try man.addDepFilePost(std.fs.cwd(), dep_output_file.generated_file.getPath());
 
+    const digest = man.final();
+
+    const any_output = output_placeholders.items.len > 0 or
+        self.captured_stdout != null or self.captured_stderr != null;
+
+    // Rename into place
+    if (any_output) {
+        const o_sub_path = "o" ++ fs.path.sep_str ++ &digest;
+        b.cache_root.handle.makePath(o_sub_path) catch |err| {
+            return step.fail("unable to make path '{}{s}': {s}", .{
+                b.cache_root, o_sub_path, @errorName(err),
+            });
+        };
+
+        b.cache_root.handle.rename(tmp_dir_path, o_sub_path) catch |err| {
+            if (err == error.PathAlreadyExists) {
+                b.cache_root.handle.deleteTree(tmp_dir_path) catch |del_err| {
+                    return step.fail("unable to remove temp dir '{}'{s}: {s}", .{
+                        b.cache_root,
+                        tmp_dir_path,
+                        @errorName(del_err),
+                    });
+                };
```

Comment:
The intent of this logic seems to be that if the destination directory already exists, then the work is already done due to a race condition. That is not the case. If the directory already exists it is because of a crash or power loss in between the rename and the `writeManifest()` call below. If the destination directory already exists at this point then the *destination* must be deleted so that the rename can continue to take place.